### PR TITLE
[core][autoscaler][v2] Reconciler - Cloud provider reconciler [3/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR implements the reconcilation logic for syncing states from the cloud provider. 

Depends on https://github.com/ray-project/ray/pull/42902

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 1]
```diff
@@ -163,3 +191,237 @@ def _step_next(
...
+            return
+        reply = instance_manager.update_instance_manager_state(
+            request=UpdateInstanceManagerStateRequest(
+                expected_version=version,
+                updates=list(updates.values()),
+            )
+        )
+        assert (
+            reply.status.code == StatusCode.OK
+        ), f"Failed to update instance manager: {reply}"
```
Comment (scv119) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:427]: hmm would this always be OK even if version conflicts?

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 2]
```diff
@@ -163,3 +191,237 @@ def _step_next(
...
+                continue
+
+            logger.debug("Updating {}".format(MessageToDict(update_event)))
+            updates[instance.instance_id] = update_event
+
+        # Update the instance manager for the events.
+        Reconciler._update_instance_manager(instance_manager, updates, version)
+
+    @staticmethod
+    def _try_or_fail_allocation(
```
Comment (scv119) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:249]: nit: _try_resolve_pending_allocation

**Review by scv119 — APPROVED** [review]
looks good to me, delta some question regarding the assert

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 3]
```diff
@@ -163,3 +191,237 @@ def _step_next(
...
+        return None
+
+    @staticmethod
+    def _handle_cloud_instance_terminated(
+        instance_manager: InstanceManager,
+        non_terminated_cloud_instances: Dict[CloudInstanceId, CloudInstance],
+    ):
+        """
+        For any IM (instance manager) instance with a cloud node id, if the mapped
+        cloud instance is no longer running, transition the instance to TERMINATED.
```
Comment (scv119) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:306]: hope the cloud provider is at least sequential consistent.
if it turns out our expectation is too high, we might give those disappeared node a grace period

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/reconciler.py:306]: good point - let me add a comment to the cloud provider.

**Review by jjyao — APPROVED** [review]

**Pull request merged**